packages/web/src/common/store/pages/audio-rewards/sagas.ts

```diff
@@ -175,8 +188,139 @@ const getClaimingConfig = (
   }
 }
 
+function* waitForOptimisticChallengeToComplete({
+  challengeId,
+  completionPollFrequency,
+  completionPollTimeout
+}: {
+  challengeId: ChallengeRewardID
+  completionPollFrequency?: number
+  completionPollTimeout?: number
+}) {
+  const challenge = yield* select(getUserChallenge, {
+    challengeId
+  })
+  if (challenge.challenge_type !== 'aggregate' && !challenge.is_complete) {
+    console.info('Waiting for challenge completion...')
+    const raceResult: { isComplete?: boolean } = yield* race({
+      isComplete: call(
+        waitForValue,
+        getUserChallenge,
+        { challengeId },
+        (challenge: UserChallenge) => challenge.is_complete
+      ),
+      poll: call(pollUserChallenges, completionPollFrequency ?? 1000),
+      timeout: delay(completionPollTimeout ?? 10000)
+    })
+    if (!raceResult.isComplete) {
+      console.warn(
+        'Challenge still not marked as completed on DN. Attempting attestations anyway, but may fail...'
+      )
+    }
+  }
+}
+
+type ErrorResult = {
+  error: unknown
+  specifier: string
+  amount: number
+}
+
 function* claimChallengeRewardAsync(
   action: ReturnType<typeof claimChallengeReward>
+) {
+  const remoteConfigInstance = yield* getContext('remoteConfigInstance')
+  const env = yield* getContext('env')
+  const audiusSdk = yield* getContext('audiusSdk')
+  const sdk = yield* call(audiusSdk)
+
+  const { claim } = action.payload
+  const { specifiers, challengeId } = claim
+
+  const { completionPollFrequency, completionPollTimeout } = getClaimingConfig(
+    remoteConfigInstance,
+    env
+  )
+
+  yield* call(waitForOptimisticChallengeToComplete, {
+    challengeId,
+    completionPollFrequency,
+    completionPollTimeout
+  })
+
+  const decodedUserId = yield* select(getUserId)
```

Comment:
We probably should decide which one of these is the canonical one. I think we have three different selectors for "current user id", and for manager mode, we probably want everything using a single selector.